### ISSUE TITLE
Unable to use the pool with Hive2 driver

### ISSUE THREAD

**Issue by mwegrz** [brettwooldridge/HikariCP#369]
Hi Brett,

I'm trying to use the pool with Hive2 JDBC driver. The implementation of the driver does not seem to be aware of the concept of transactions. Namely (as of version 1.2.1):

- `Connection.setAutoCommit(boolean)` can only be set to `false` (throws an exception otherwise) [link](https://github.com/apache/hive/blob/release-1.2.1/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L1197),
- `Connection.rollback()` is unsupported (throws an exception) [link](https://github.com/apache/hive/blob/release-1.2.1/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L1173),
- `Connection.commit()` is unsupported (throws an exception) [link](https://github.com/apache/hive/blob/release-1.2.1/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L740).

This virtually renders the pool unsuable because when `auto-commit` is set to `false` (the only viable setting) the pool tries to rollback the connection in `Connection.close()` which is unsupported in Hive2.

This might be an issue with Hive2 driver but there might be other drivers with similar partly-implemented API.
The resolution to this in HikariCP could be having the `auto-commit` undefined. In such a case the pool would not call `Connection.setAutoCommit` at all (simillar to [`Connection.setReadOnly`](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-2.4.0/src/main/java/com/zaxxer/hikari/pool/PoolElf.java#L171) ) and disable rollback when closing the connection. 

Another option I could think of is to have a setting for disabling full transaction tracking.

I can provide a pull request with this enhancement. Please let me know what do you think.

Cheers,
Michał

**Comment by brettwooldridge** [comment]
The fact that auto-commit can only be set to ``false`` and yet rollback/commit throw exceptions is very strange.  It is a semantic mismatch.  Auto-commit = ``true`` implies that commit and rollback are pointless, whereas ``false`` would imply that either a commit or rollback is required.

I haven't traced through the HikariCP code to know for sure, but my suspicion is that if their driver supported a value of ``true``, HikariCP would never attempt to rollback.

**Comment by brettwooldridge** [comment]
@mwegrz The ``readOnly`` exception added for VoltDB set a bad precedent, and in truth should be removed from HikariCP.  Both VoltDB and Hive2 are open-source projects and the issue should be taken up there.  A JDBC-compliant driver is not allowed to throw ``SQLFeatureNotSupportedException`` unless the API documentation or JDBC specification explicitly allows it.  That is not the case for either ``Connection.setReadOnly()`` or ``Connection.setAutoCommit()``.

The exception documentation for ``setAutoCommit()`` states that a SQLException is thrown...
```
if a database access error occurs, setAutoCommit(true) is called while participating in a distributed transaction, or this method is called on a closed connection
```
The correct behavior for the driver (both Hive2 and VoltDB) is to silently ignore the call.

As an additional note, the auto-commit case is not nearly as simple as the Connection.setReadOnly case.  HikatiCP is keeping track of transaction state as well as connection state, so the code impact is much larger.  While it is true that it's just code, and can be made to do anything, there are real impacts to performance, testability, readability, and maintainability when we start layering on this kind of exceptional logic.

**Comment by mwegrz** [comment]
Thanks for comments. I fully agree that driver implementations should be compliant with the JDBC spec. For people having same issues, there are at least 2 tickets open for this specific case of strange behavior of `Connection.setAutoCommit()`: [HIVE-6705] (https://issues.apache.org/jira/browse/HIVE-6705), [HIVE-11293] (https://issues.apache.org/jira/browse/HIVE-11293). The fix needs to be done in Hive2's `Connection` implementation.

Nevertheless, I've implemented the discussed change in HikariCP and will use it until Hive2 has been fixed. The patch (based on 2.4.0 release) is available [here] (https://github.com/mwegrz/HikariCP/commit/e19c5e3d18c628ef38f03632577b7c0d8a1fd129).
